New frontend web client written using Rust and Yew.rs

chris-dietz opened the pull request:
Adds a new web client runner built from the ground up using rust and yew.rs. 
Part of #261 

Right now web client must be manually built and launched separate to launching the backend. A followup PR will add a new flag to tell runner to use the new web client. Both the old and new client will be supported until the new front end has been thoroughly tested and has feature parity.

ia0 commented on crates/runner-host/crates/web-client/src/app.rs:
```diff
@@ -0,0 +1,70 @@
+// Copyright 2023 Google LLC
```
```suggestion
// Copyright 2024 Google LLC
```

chris-dietz replied:
Fixed

ia0 commented on crates/runner-host/crates/web-client/src/board.rs:
```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 Google LLC
```
(same in all new files)
```suggestion
// Copyright 2024 Google LLC
```

chris-dietz replied:
Fixed

ia0 commented on crates/runner-host/crates/web-client/src/board.rs:
```diff
@@ -0,0 +1,95 @@
+#[function_component]
+pub fn Board(Props { command_state, on_board_ready, on_event }: &Props) -> Html {
+    let board_config = use_state_eq(|| None);
+    let enabled_leds = use_set(HashSet::new());
```
This looks weird to me that Yew forces the state of sub-components to be in the parent. Either there is a better way using function components, or we have to use struct components in our case. The `Component` interface looks more appropriate to message passing to me (using the `update` function and thus matching the messages we get on the websocket), than those hooks.

Did you try implementing `Component` instead of using `function_component` macro? I would expect a design where the app would start a websocket and its update function would be called each time there is a message (not sure how that would be done), and then the app would dispatch the message to the appropriate sub-component.

chris-dietz replied:
I reworked this to store the state of the led inside of the led component and just pass it the command state so it can update itself. I think it should be possible to do that with arbitrary function components with a bit of thinking so I think we should be able to avoid stuct components. 

Let me know what you think of the new version. If that is the type of thing you were going for.

ia0 replied:
Yes, this looks good. Thanks!

ia0 commented on crates/runner-host/crates/web-client/src/hooks/use_runner_connection.rs:
```diff
@@ -0,0 +1,63 @@
+#[hook]
+pub fn use_runner_connection(backend_address: String) -> UseRunnerConnectionHandle {
+    let ws = use_websocket(backend_address.clone());
+    let command_state = use_state_eq(|| None);
```
Why `use_state_eq` instead of `use_state`? It seems to me that this would deduplicate identical messages.

ia0 commented on crates/runner-host/crates/web-common/src/lib.rs:
```diff
@@ -51,6 +51,10 @@ pub enum Command {
+
+    // Input from user to console.
+    #[serde(rename_all = "camelCase")]
+    Input { message: String },
```
We don't support console input yet (see #236), but I guess you added this for testing purposes. We could indeed just print those in the host runner with a warning that it was ignored.

ia0 replied:
I still don't understand why you added input support. Where do you need it? I would suggest simply removing console input in the Web UI completely. This is misleading since users can't input anything from the Web UI. The current UI is already misleading which is probably why you added it. In the current UI I don't think anything is sent over web-socket. This is purely something on JS side.

So I would actually suggest removing this to avoid further misunderstanding of what is supported.

chris-dietz replied:
The reason was I wanted to exactly match the behavior of the current web client. 
Currently if you observe the websocket activity of the current client on console input it sends the following json to the backend:
```
{
type: "input"
message: "<Message contents>"
}
```

I'd have no problem removing it for now though. Just with the caveat that it will now behave differently than the current web client. 

If that sounds good to you I'll go ahead and remove it.

ia0 replied:
Yes, I think it's a bug of the current UI, not a feature. We should remove this.

chris-dietz replied:
Alright sg. I have removed sending console input. I still left most of the plumbing for it in place though so it should be quite easy to add sending console input to the backend in the future.

ia0 reviewed (commented):
Thanks! I didn't review everything yet because this is quite big. Also I have a high-level concern which is that Yew seems to force putting the state of a component in its parent through the notion of properties. I find this quite weird, but maybe that's how it's supposed to be? It seems that implementing `Component` manually would fix this because we have control over the `Message` associated type. But then, maybe this will make other things complicated. I would be curious to know if you tried implementing `Component` explicitly, such that each component has its own state and the properties are only used for configuration by the parent, while the state is modified through message passing. This seems to better match the fact that the app is driven by a websocket which is about message passing.

chris-dietz commented on crates/runner-host/crates/web-client/README.md:
````diff
@@ -0,0 +1,24 @@
+### Running
+
+```bash
+trunk serve
````
Yeah I believe something like that should be possible. Once truck build has been run it can be served just like a static page. I'll look into that for when we I add a flag to the runner

chris-dietz commented on crates/runner-host/crates/web-client/.gitignore:
```diff
@@ -0,0 +1,2 @@
+/dist/
+/target/
```
Removed /target

I still need /dist for trunk build artifacts.

chris-dietz commented on crates/runner-host/crates/web-client/README.md:
```diff
@@ -0,0 +1,24 @@
+# Wasefire Web Client
+
+This is a fairly minimal web app for interacting with the runner that's built with Yew and [Trunk].
```
Added links to both trunk and yew.

chris-dietz commented on crates/runner-host/crates/web-client/index.html:
Done. Except for index.html. That has to live in the root for trunk I believe.

chris-dietz commented on crates/runner-host/crates/web-client/components/button_pressed.svg:
I think it would certainly be possible. When I was implementing it I initially tried to replicate the exact way the old client did things but I ran into enough road blocks that I decided it wasn't going to be worth it for the moment.  The main issue is it was going to be difficult to use the svg.js library from within yew and being a non svg expert I decided it wasn't worth the effort right now to figure out how to replicate the exact way the old client did things. 

I think it's something I could look into in the future but imo I don't think it's necessary at the moment just to get to basic functionality. If you disagree strongly though I can look into doing that now. 

Also in terms of using the assets from web-server/public I think it makes more sense for all the webclient assets to live within the web-client crate. And we can just delete /web-server/public after the new client is ready to replace it.

ia0 commented on crates/runner-host/crates/web-client/src/console.rs:
```diff
@@ -0,0 +1,83 @@
+use web_sys::HtmlInputElement;
+use yew::prelude::*;
+use yew_hooks::prelude::*;
+#[derive(Properties, PartialEq)]
```
same here, let's have an empty line between items

ia0 commented on crates/runner-host/crates/web-client/src/console.rs:
```diff
@@ -0,0 +1,83 @@
+            let history = history.clone();
+            move |command_state| {
+                if let Some(command) = &**command_state {
+                    info!("Command: {:?} ", command);
```
is the trailing space intended?
```suggestion
                    info!("Command: {:?}", command);
```

ia0 commented on crates/runner-host/crates/web-client/src/console.rs:
```diff
@@ -0,0 +1,83 @@
+        });
+    }
+
+    html! {
+        <div id={id.to_string()} class={"console"}>
+            <p>
+                <b>{ "Log history: " }</b>
+            </p>
+            {
+                for history.current().iter().map(|message| {
+                    html! {
+                        <p>{ message }</p>
+                    }
+                })
+            }
+            <form onsubmit={onsubmit}>
+                <input ref={console_ref} type="text" id="consolein" />
+                <input  type="submit" value="Send"/>
+            </form>
+        </div>
+    }
```
The style of this macro looks much better. I like:
- The indentation matching the HTML AST.
- The extra space withing curly braces to strengten switching from HTML to Rust. I would do it also for parameters like `id` and `class`.

ia0 commented on crates/runner-host/crates/web-client/src/console.rs:
```diff
@@ -0,0 +1,83 @@
+            <p>
+                <b>{ "Log history: " }</b>
+            </p>
+            {
+                for history.current().iter().map(|message| {
+                    html! {
+                        <p>{ message }</p>
+                    }
+                })
+            }
```
nit: this could be a single line if you want (macros can be called with parentheses)
```suggestion
            { for history.current().iter().map(|message| html!(<p>{ message }</p>)) }
```

ia0 commented on crates/runner-host/crates/web-client/src/hooks.rs:
```diff
@@ -0,0 +1,14 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+pub mod use_runner_connection;
```
nit: empty line, but anyway, let's remove this module completely and inline its content in the hooks module directly

ia0 commented on crates/runner-host/crates/web-client/src/hooks/use_runner_connection.rs:
```diff
@@ -0,0 +1,65 @@
+    let ws = use_websocket(backend_address.clone());
+    let command_state = use_state(|| None);
+
+    {
+        let ws = ws.clone();
+        // Receive message by depending on `ws.message`.
+        use_effect_with(ws.message, {
```
Same here, and it's actually even worse because you only need to clone the message, not the whole websocket.
```suggestion
    // Receive message by depending on `ws.message`.
    use_effect_with(ws.message.clone(), {
```

ia0 commented on crates/runner-host/crates/web-client/src/hooks/use_runner_connection.rs:
```diff
@@ -0,0 +1,65 @@
+                if let Some(message) = &**message {
+                    info!("Message: {message}");
+                    match serde_json::from_str::<Command>(message) {
+                        Ok(command) => {
+                            command_state.set(Some(command));
+                        }
```
nit: by removing the semi-colon, this will end up on one line
```suggestion
                        Ok(command) => command_state.set(Some(command)),
```

ia0 commented on crates/runner-host/crates/web-client/src/main.rs:
```diff
@@ -0,0 +1,25 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod app;
+mod board;
+mod board_components;
+mod console;
+mod hooks;
+use app::App;
```
nit: use items go before mod items (that may look weird but that's how Rust does it)
```suggestion
use app::App;

mod app;
mod board;
mod board_components;
mod console;
mod hooks;
```

ia0 commented on crates/runner-host/crates/web-client/test.sh:
```diff
@@ -0,0 +1,23 @@
+test_helper
+
+cargo test
+cargo check --features=wasefire-logger/log
```
We can actually remove this line since we don't use wasefire-logger
```suggestion
```

ia0 reviewed (commented):
Thanks! I think the overall design is there. Now it's mostly about style (except for the console input which I think should be removed).

chris-dietz commented on crates/runner-host/crates/web-client/Cargo.toml:
```diff
@@ -0,0 +1,27 @@
+serde_json = "1.0.111"
+wasefire-logger = { path = "../../../logger" }
+wasm-bindgen = "0.2.92"
+wasm-bindgen-futures = "0.4.42"
```
I'll remove these but 
Is there a lint for unused dependencies I could run? Would be good to know for future reference.

ia0 replied:
Yes, there's one. I'm trying to find a way to have it consistently through the whole repo. For local development you can add this to your `Cargo.toml`:

```toml
[lints.rust]
unused-crate-dependencies = "warn"
```

ia0 replied:
I've merged the lints into main. You can pull and run `./scripts/sync.sh` to sync the lints.

chris-dietz replied:
Awesome, tysm.

chris-dietz commented on crates/runner-host/crates/web-client/src/app.rs:
```diff
@@ -0,0 +1,69 @@
+
+use crate::board::Board;
+use crate::console::Console;
+use crate::hooks::use_runner_connection::use_runner_connection;
```
Umm sure. I think it's likely that we will add more hooks in the future though so having a directory for them with one hook per file made sense to me. But I don't feel that strongly so if you want to keep all of the hooks in one file / wait to refactor in the future when there's more I'm good with that.

ia0 replied:
Yes, let's do that. The web UI is almost feature complete, so I don't expect much more hooks. Besides, hooks don't seem to need helper functions, so they naturally don't have conflicting names and can live in the same module.

chris-dietz commented on crates/runner-host/crates/web-client/src/board.rs:
```diff
@@ -0,0 +1,82 @@
+                }
+
+            }).collect::<Html>()
+            }else{
```
Oh I see, I installed a formatter that was supposed to be yew aware but it seems like it's not that consistent with things.

ia0 commented on crates/runner-host/crates/web-client/src/app.rs:
```diff
@@ -0,0 +1,73 @@
+
+    let on_board_ready = Callback::from({
+        let runner_connection = runner_connection.clone();
+        move |()| {
+            runner_connection.send_board_ready();
+        }
```
nit: removing the semi-colon would simplify this closure.
```suggestion
        move |()| runner_connection.send_board_ready()
```

ia0 commented on crates/runner-host/crates/web-client/src/app.rs:
```diff
@@ -0,0 +1,73 @@
+
+    let send_event_callback = Callback::from({
+        let runner_connection = runner_connection.clone();
+        move |event: Event| {
+            runner_connection.send_event(event);
+        }
```
same here
```suggestion
        move |event: Event| runner_connection.send_event(event)
```

ia0 commented on crates/runner-host/crates/web-client/src/app.rs:
```diff
@@ -0,0 +1,73 @@
+            match command {
+                Command::Connected => info!("Connected to runner"),
+                Command::Disconnected => warn!("Disconnected from runner"),
+                _ => {} // Command for other component so ignoring.
```
nit
```suggestion
                _ => (), // Command for other component so ignoring.
```

ia0 commented on crates/runner-host/crates/web-client/src/board.rs:
```diff
@@ -0,0 +1,79 @@
+        let on_board_ready = on_board_ready.clone();
+        move |command_state| {
+            if let Some(Command::BoardConfig { components }) = &**command_state {
+                info!("Board config: {:?}", components);
```
```suggestion
                info!("Board config: {components:?}");
```

ia0 commented on crates/runner-host/crates/web-client/src/app.rs:
```diff
@@ -0,0 +1,73 @@
+    use_effect_with(runner_connection.command_state.clone(), move |command_state| {
+        if let Some(command) = &**command_state {
+            info!("Command: {command:?}");
+
```
nit
```suggestion
```

ia0 commented on crates/runner-host/crates/web-client/src/board.rs:
```diff
@@ -0,0 +1,79 @@
+                board_config.set(Some(components.clone()));
+                on_board_ready.emit(());
+            }
+
```
nit
```suggestion
```

ia0 commented on crates/runner-host/crates/web-client/src/board.rs:
```diff
@@ -0,0 +1,79 @@
+        }
+    });
+
+    return html! {
+
+
+        <div class="board">
+        {
+            if let Some(board_config) = &*board_config {
+                board_config.iter().map(|component| html! {
+                    match component {
+                        Component::Button{id} => {
+                            html! {
+                                <Button id={id} on_event={on_event} />
+                            }
+                        }
+                        Component::MonochromeLed{id} => {
+                            html! {
+                                <LED id={id} command_state={command_state.clone()}/>
+                            }
+                        }
+                    }
+
+                }).collect::<Html>()
+            } else {
+                html! {
+                    <div>{"Waiting for backend connection..."}</div>
+                }
+            }
+
+
+        }
+        </div>
+    };
```
```suggestion
    return html! {
        <div class="board">{
            if let Some(board_config) = &*board_config {
                board_config.iter().map(|component| match component {
                    Component::Button{id} => html!(<Button id={id} on_event={on_event} />),
                    Component::MonochromeLed{id} => html! {
                        <LED id={id} command_state={command_state.clone()}/>
                    }
                }).collect::<Html>()
            } else {
                html!(<div>{"Waiting for backend connection..."}</div>)
            }
        }</div>
    };
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/button.rs:
```diff
@@ -0,0 +1,78 @@
+#[function_component]
+pub fn Button(Props { id, on_event }: &Props) -> Html {
+    let pressed = use_state(|| false);
+    let press = {
```
This layer of braces shouldn't be needed. See unpress below.

ia0 commented on crates/runner-host/crates/web-client/src/board_components/button.rs:
```diff
@@ -0,0 +1,78 @@
+        let id = *id;
+        move |_| {
+            info!("Unpressed button id: {id}");
+            if *pressed {
```
Is this `if` really needed?

ia0 commented on crates/runner-host/crates/web-client/src/board_components/button.rs:
```diff
@@ -0,0 +1,78 @@
+            onmousedown={press}
+            onmouseup={unpress.clone()}
+            onmouseleave={unpress.clone()}
+            class= {"button"}
```
```suggestion
            class={"button"}
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/button.rs:
```diff
@@ -0,0 +1,78 @@
+                style={if !*pressed { "" } else { "display: none;" }}
+            />
+        </div>
+
```
```suggestion
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/led.rs:
```diff
@@ -0,0 +1,60 @@
+            class="led" type="image/svg+xml"
+            data="components/monochrome_led_on.svg"
+            style={if *lit { "" } else { "display: none;" }}
+        >
+        </object>
```
If I remember correctly, Yew documentation said that they support self-closing tags even if they should not.
```suggestion
        />
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/led.rs:
```diff
@@ -0,0 +1,60 @@
+    });
+
+    return html! {
+    <div class="monochrome_led" id={id.to_string()}>
```
In other places we have indentation within the html block. Let's also indent here.

ia0 commented on crates/runner-host/crates/web-client/src/console.rs:
```diff
@@ -0,0 +1,78 @@
+                console_ref.get().unwrap().value_of().dyn_into().unwrap();
+            let value = input_form.value();
+            info!("sending console message: {value}");
+            history.push(format!("[send]: {}", value));
```
```suggestion
            history.push(format!("[send]: {value}"));
```

ia0 commented on crates/runner-host/crates/web-client/src/console.rs:
```diff
@@ -0,0 +1,78 @@
+        }
+    });
+
+    {
+        let command_state = command_state.clone();
+        use_effect_with(command_state, {
```
The clone can be inlined and the curly braces removed

ia0 commented on crates/runner-host/crates/web-client/src/console.rs:
```diff
@@ -0,0 +1,78 @@
+                if let Some(command) = &**command_state {
+                    info!("Command: {:?}", command);
+                    if let Command::Log { message } = command {
+                        history.push(format!("[recv]: {}", message));
```
```suggestion
                        history.push(format!("[recv]: {message}"));
```

ia0 commented on crates/runner-host/crates/web-client/src/console.rs:
```diff
@@ -0,0 +1,78 @@
+            let history = history.clone();
+            move |command_state| {
+                if let Some(command) = &**command_state {
+                    info!("Command: {:?}", command);
```
```suggestion
                    info!("Command: {command:?}");
```

ia0 commented on crates/runner-host/crates/web-client/src/console.rs:
```diff
@@ -0,0 +1,78 @@
+
+    html! {
+        <div id={id.to_string()} class={"console"}>
+            <p>
+                <b>{ "Log history: " }</b>
+            </p>
```
```suggestion
            <p><b>{ "Log history: " }</b></p>
```

ia0 commented on crates/runner-host/crates/web-client/src/hooks/use_runner_connection.rs:
```diff
@@ -0,0 +1,62 @@
+    let ws = use_websocket(backend_address.clone());
+    let command_state = use_state(|| None);
+
+    {
```
No need for those braces

ia0 commented on crates/runner-host/crates/web-client/src/hooks/use_runner_connection.rs:
```diff
@@ -0,0 +1,62 @@
+            }
+        });
+    }
+    return UseRunnerConnectionHandle { ws, command_state };
```
no need for return
```suggestion
    UseRunnerConnectionHandle { ws, command_state }
```

ia0 reviewed (commented):
Thanks! This looks much better now. There's still quite a lot of clones and weird things but I guess that's how Yew is supposed to work. And we could always improve later as we learn Yew better.

ia0 commented on crates/runner-host/crates/web-client/test.sh:
```diff
@@ -0,0 +1,22 @@
+
+test_helper
+
+cargo test
```
```suggestion
cargo test --bin=web-client
```

chris-dietz replied:
Ah, I see thanks. Was trying to figure out why that was failing xD

ia0 reviewed (commented):
I guess this is not yet ready for review.

I've created a `dev/web-ui` branch (so that we can merge intermediate PRs) and fixed the `test.sh` which format changed since last `main`.

ia0 commented:
I guess you are working incrementally or did not push your changes. Feel free to click the "Re-request review" button once the PR is ready for review. Thanks!

chris-dietz commented:
> I guess you are working incrementally or did not push your changes. Feel free to click the "Re-request review" button once the PR is ready for review. Thanks!

Will do. I think the confusion is I seemed to have messed something up and lost some changes that I made and I hadn't noticed. Something with committing with the web UI and then making and pushing local changes resulted in me losing some work. I think I have redone all of my changes now. 

I'm going to do another pass and then I'll request a review again. Sorry about that.

chris-dietz commented:
Alright I think I have incorporated all the feedback and I also went back and cleaned up the commit history. So I think it should be ready to be re-reviewed now. 

Sorry for any confusion. I'll do a better job of waiting to push until I'm ready for things to be reviewed in the future.

ia0 commented on crates/runner-host/crates/web-client/src/board.rs:
```diff
@@ -0,0 +1,61 @@
+                board_config.iter().map(|component| match component {
+                    Component::Button{id} => html!(<Button id={id} on_event={on_event} />),
+                    Component::MonochromeLed{id} => html! {
+                        <LED id={id} command_state={command_state.clone()}/>
```
nit
```suggestion
                        <LED id={id} command_state={command_state.clone()} />
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/button.rs:
```diff
@@ -0,0 +1,78 @@
+
+#[function_component]
+pub fn Button(Props { id, on_event }: &Props) -> Html {
+    let pressed = use_state(|| false);
```
nit: usize is copy, so we can dereference it only once.
```suggestion
    let id = *id;
    let pressed = use_state(|| false);
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/button.rs:
```diff
@@ -0,0 +1,78 @@
+    let press = Callback::from({
+        let pressed = pressed.clone();
+        let on_event = on_event.clone();
+        let id = *id;
```
```suggestion
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/button.rs:
```diff
@@ -0,0 +1,78 @@
+    let unpress = Callback::from({
+        let pressed = pressed.clone();
+        let on_event = on_event.clone();
+        let id = *id;
```
```suggestion
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/button.rs:
```diff
@@ -0,0 +1,78 @@
+        let id = *id;
+        move |_| {
+            info!("Unpressed button id: {id}");
+            // Necessary because it may also be triggered when the mouse
+            // stops hovering over the button.
```
nit: column width is 100 for comments too (I could try to enforce this with rustfmt)
```suggestion
            // Necessary because it may also be triggered when the mouse stops hovering over the
            // button.
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/led.rs:
```diff
@@ -0,0 +1,59 @@
+
+#[function_component]
+pub fn LED(Props { id, command_state }: &Props) -> Html {
+    let lit = use_state(|| false);
```
```suggestion
    let id = *id;
    let lit = use_state(|| false);
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/led.rs:
```diff
@@ -0,0 +1,59 @@
+    let command_state = command_state.clone();
+    use_effect_with(command_state, {
+        let lit = lit.clone();
+        let id = *id;
```
```suggestion
```

ia0 commented on crates/runner-host/crates/web-client/src/board_components/led.rs:
```diff
@@ -0,0 +1,59 @@
+                data="components/monochrome_led_on.svg"
+                style={if *lit { "" } else { "display: none;" }}
+            />
+
```
```suggestion
```

ia0 commented on crates/runner-host/crates/web-client/src/hooks.rs:
```diff
@@ -0,0 +1,58 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use log::{info, warn};
```
```suggestion

use log::{info, warn};
```

ia0 reviewed (commented):
Thanks! I'll apply the suggestions and merge. Then I'll merge main into dev/web-ui, so you'll have to pull dev/web-ui and work on that branch. In particular, now that we have a branch for this work, you can remove the automatic opening of the old UI and do any change convenient to you until we can replace the old UI, at which point we can merge this branch into main.

ia0 commented:
> I'll apply the suggestions and merge.

I spoke too fast. Looks like this PR doesn't have "allow edits by maintainers", so I can't. I'll wait until you apply the suggestions then.

chris-dietz commented:
> > I'll apply the suggestions and merge.
> 
> I spoke too fast. Looks like this PR doesn't have "allow edits by maintainers", so I can't. I'll wait until you apply the suggestions then.

I applied the suggested changes. So should be good to merge I think. :)

ia0 reviewed (approved):
Thanks!

ia0 commented:
I've merge main into dev/web-ui. You'll need to continue working from that branch.
